[Opt Overlap] Clean up code in _OptimizerHookState

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71620
* #71608
* #71607
* #71606
* #71605
* #71604
* #71603
* #71602

Remove from_functional_optim and make it the default constructor since
that is the only way _OptimizerHookState is now being built. Also, no longer
need to expose create_functional_optim helper function

Differential Revision: [D33700593](https://our.internmc.facebook.com/intern/diff/D33700593/)

rohan-varma committed:
[Opt Overlap] Clean up code in _OptimizerHookState

Remove from_functional_optim and make it the default constructor since
that is the only way _OptimizerHookState is now being built. Also, no longer
need to expose create_functional_optim helper function

Differential Revision: [D33700593](https://our.internmc.facebook.com/intern/diff/D33700593/)

[ghstack-poisoned]

cbalioglu reviewed (approved):
Nice! Please ignore my comments about this issue in my reviews for previous PRs in this stack.

cbalioglu commented:
Same as the previous PR. Please verify the failing distributed tests before shipping this.

rohan-varma committed:
Update on "[Opt Overlap] Clean up code in _OptimizerHookState"


Remove from_functional_optim and make it the default constructor since
that is the only way _OptimizerHookState is now being built. Also, no longer
need to expose create_functional_optim helper function

Differential Revision: [D33700593](https://our.internmc.facebook.com/intern/diff/D33700593/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Clean up code in _OptimizerHookState"


Remove from_functional_optim and make it the default constructor since
that is the only way _OptimizerHookState is now being built. Also, no longer
need to expose create_functional_optim helper function

Differential Revision: [D33700593](https://our.internmc.facebook.com/intern/diff/D33700593/)

[ghstack-poisoned]

rohan-varma commented:
CI failure is unrelated: https://app.circleci.com/pipelines/github/pytorch/pytorch/445245/workflows/cbcc6b7b-c346-4efe-831b-726ebb65ad20/jobs/16965362
